src/scheduler/SnapshotServer.cpp

```diff
@@ -1,40 +1,47 @@
-#include <faabric/scheduler/MpiWorldRegistry.h>
 #include <faabric/scheduler/SnapshotServer.h>
 #include <faabric/snapshot/SnapshotRegistry.h>
 #include <faabric/state/State.h>
-#include <faabric/util/config.h>
 #include <faabric/util/func.h>
 #include <faabric/util/logging.h>
 
-#include <faabric/rpc/macros.h>
-#include <grpcpp/grpcpp.h>
-
 namespace faabric::scheduler {
 SnapshotServer::SnapshotServer()
-  : RPCServer(DEFAULT_RPC_HOST, SNAPSHOT_RPC_PORT)
+  : faabric::transport::MessageEndpointServer(DEFAULT_SNAPSHOT_HOST,
+                                              SNAPSHOT_PORT)
 {}
 
-void SnapshotServer::doStart(const std::string& serverAddr)
+void SnapshotServer::doRecv(faabric::transport::Message& header,
+                            faabric::transport::Message& body)
 {
-    // Build the server
-    ServerBuilder builder;
-    builder.AddListeningPort(serverAddr, InsecureServerCredentials());
-    builder.RegisterService(this);
-
-    // Start it
-    server = builder.BuildAndStart();
-    faabric::util::getLogger()->info("Snapshot server listening on {}",
-                                     serverAddr);
-
-    server->Wait();
+    assert(header.size() == sizeof(uint8_t));
+    uint8_t call = static_cast<uint8_t>(*header.data());
+    switch (call) {
+        case faabric::scheduler::SnapshotCalls::PushSnapshot:
+            this->recvPushSnapshot(body);
+            break;
+        case faabric::scheduler::SnapshotCalls::DeleteSnapshot:
+            this->recvDeleteSnapshot(body);
+            break;
+        default:
+            throw std::runtime_error(
+              fmt::format("Unrecognized call header: {}", call));
+    }
 }
 
-Status SnapshotServer::PushSnapshot(
-  ServerContext* context,
-  const flatbuffers::grpc::Message<SnapshotPushRequest>* request,
-  flatbuffers::grpc::Message<SnapshotPushResponse>* response)
+void SnapshotServer::recvPushSnapshot(faabric::transport::Message& msg)
 {
-    const SnapshotPushRequest* r = request->GetRoot();
+    // Persist message to ensure contents are not cleared
+    // Note that these are currently not free-d, see TODO in next function
+    msg.persist();
```

Comment:
The memory `mmap`-ed for the snapshot push request needs _not_ to be freed when the `transport::Messge` goes out of scope, so we mark it as persistent. We though need to make sure we `free` the underlying pointer aftterwards.